pkg/nimble/autoadv/nimble_autoadv.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2020 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     pkg_nimble_autoadv
+ *
+ * @{
+ *
+ * @file
+ *
+ * @author      Hendrik van Essen <hendrik.ve@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include "nimble_riot.h"
+
+#include "host/ble_hs.h"
+#include "host/ble_gap.h"
+#include "net/bluetil/ad.h"
+
+#include "nimble_autoadv.h"
+
+static struct ble_gap_adv_params _advp;
+
+static int32_t _adv_duration = BLE_HS_FOREVER;
+
+static uint8_t buf[BLE_HS_ADV_MAX_SZ];
+static bluetil_ad_t _ad;
+
+static int _gap_event_cb(struct ble_gap_event *event, void *arg);
+
+static ble_gap_event_fn *_gap_cb = &_gap_event_cb;
+static void *_gap_cb_arg = NULL;
+
+void nimble_autoadv_start(void);
+
+static int _gap_event_cb(struct ble_gap_event *event, void *arg)
+{
+    (void) arg;
+
+    switch (event->type) {
+
+        case BLE_GAP_EVENT_CONNECT:
+            if (event->connect.status != 0) {
+                // failed, ensure advertising is restarted
+                nimble_autoadv_start();
+            }
+            break;
+
+        case BLE_GAP_EVENT_DISCONNECT:
+            nimble_autoadv_start();
+            break;
+    }
+
+    return 0;
+}
+
+void nimble_autoadv_init(void)
+{
+    memset(&_advp, 0, sizeof _advp);
+    _advp.conn_mode = BLE_GAP_CONN_MODE_UND;
+    _advp.disc_mode = BLE_GAP_DISC_MODE_GEN;
+
+    int rc = 0;
+    (void) rc;
+
+    rc = bluetil_ad_init_with_flags(&_ad, buf, sizeof(buf), BLUETIL_AD_FLAGS_DEFAULT);
+    assert(rc == BLUETIL_AD_OK);
+
+    if (NIMBLE_AUTOADV_DEVICE_NAME != NULL) {
+        rc = bluetil_ad_add_name(&_ad, NIMBLE_AUTOADV_DEVICE_NAME);
+        assert(rc == BLUETIL_AD_OK);
+    }
+
+    if (NIMBLE_AUTOADV_START_MANUALLY) {
+        nimble_autoadv_start();
+    }
+}
+
+int nimble_autoadv_add_field(uint8_t type, const void *data, size_t data_len)
```

Comment:
just noticed: there is currently no way to reset the advertising data at runtime, right? I wonder if there is an easy and efficient way to add that functionality. I see two options: add a separate function like `nimble_autoadv_ad_reset()` or add something like a reset flag to this function here. I feel pretty much indifferent about this, so I let you decide whether you want to do this at all or which way would be best :-)

Reply:
I like the idea. It gives us a reset function without a code overhead, because I can simply move the initialization code in to the reset function. I squashed the previous fixups and pushed a new commit with my proposal for a reset function.

Reply:
So what I did is a complete reset. I believe you meant a reset only for the data stored in `_ad`. I don't really have a use case for a reset myself on hand, but I think if I would like to use it, than I would like to reset everything and make necessary changes afterwards.